### PR TITLE
Fog Ledger Router integration test

### LINKED ISSUE #2830: Stores with overlapping block ranges

### PR DESCRIPTION
**Pull request by awygle**
### Motivation

The most interesting functionality of the Fog Ledger Router requires some kind of load balancer to be tested. This is an integration test which tests the behavior of the FLR under real-world (ish) conditions, using localhost ports and direct DB access but otherwise fully exercising the APIs.

Closes #2830 .

### Future Work
We need a test for conflicting information coming from stores, but first we need a decision on what the correct response to that is. Since it's an inconsistency in our blockchain (kinda), it should probably throw an error.

[Soundtrack of this PR]()

### CODE REVIEW

**Review by nick-mobilecoin — CHANGES_REQUESTED** [review]
The main blockers are:
- testing [1,4] instead of [1, 5]
- Currently if an assert fails the test stays running.  Adding a drop implementation to the proxy server, which calls stop resolves this runaway test

While not a blocker I'm a bit concerned with continual addition of large integration tests like these. I know there is #3168 which doesn't seem to cover this. We also need to get things done. I spent more time than I care to admit trying to understand all the pieces necessary for these tests to execute. Only time will tell how soon and how  much we trip up on these large tests. But I can't shake the feeling we keep digging the hole deeper.

File: fog/ledger/server/tests/router_integration.rs [hunk 1]
```diff
@@ -0,0 +1,572 @@
...
+}
+
+fn temp_dir(name: &str) -> Result<PathBuf, std::env::JoinPathsError> {
+    let path: PathBuf = [std::env::temp_dir(), std::path::PathBuf::from(name)]
+        .iter()
+        .collect();
+    std::fs::create_dir_all(&path)
+        .unwrap_or_else(|_| panic!("Couldn't create temporary path {:?}", &path));
+    Ok(path)
+}
```
[the remark this answers is not in the document]
Comment (NotGyro) [reply fog/ledger/server/tests/router_integration.rs:79]: Seconding this here.

**Review by NotGyro — APPROVED** [review]
Looks good to me

File: fog/ledger/test_infra/src/lib.rs [hunk 2]
```diff
@@ -286,3 +295,67 @@ impl Ledger for MockLedger {
         unimplemented!()
     }
 }
+
+pub struct ShardProxyServer {
+    server_handle: JoinHandle<Result<(), hyper::Error>>,
+    stop_channel: oneshot::Sender<()>,
+}
+
+impl ShardProxyServer {
```
Comment (nick-mobilecoin) [inline fog/ledger/test_infra/src/lib.rs:304]: I don't think these tests are providing any assurances about the behavior of a shard with multiple stores. Thus I think this is adding bunch of complexity to the tests with little to no assurances of behavior.

These tests look like they are testing
```mermaid
flowchart  LR
A(Router) --> B(Shard 1)
B --> F[Store A]
B --> G[Store B]
B --> H[Store C]
```
but this `ShardProxyServer` is faking out the shards, so these tests are
```mermaid
flowchart  LR
A(Router) --> B(Fake Shard 1)
B --> F[Store A]
B --> G[Store B]
B --> H[Store C]
```
Generally in testing one uses a test double (fake, mock, stub, etc) at the edges of what's being tested, but this is in the middle of what's being tested.

File: fog/ledger/server/tests/router_integration.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply fog/ledger/server/tests/router_integration.rs]: was updated to use https://docs.rs/tempdir/latest/tempdir/

File: fog/ledger/server/tests/router_integration.rs [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+
+    let ledger_db_tmp =
+        temp_dir(&test_config.address.to_string()).expect("Could not get test_ledger tempdir");
+    let ledger_db_path = ledger_db_tmp.as_path();
+    let mut ledger = recreate_ledger_db(ledger_db_path);
+
+    let watcher_db_tmp =
+        temp_dir(&test_config.address.to_string()).expect("Could not make tempdir for watcher db");
+    let watcher_db_path = watcher_db_tmp.as_path();
+    let mut watcher = setup_watcher_db(watcher_db_path.to_path_buf(), logger.clone());
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply fog/ledger/server/tests/router_integration.rs]: no longer an issue due to refactor

File: fog/ledger/server/tests/router_integration.rs [hunk 4]
```diff
@@ -0,0 +1,572 @@
...
+
+        add_block_to_ledger(ledger, &recipients, &key_images, &mut rng, watcher);
+    }
+}
+
+fn create_store(
+    test_config: &StoreConfig,
+    blocks_config: &BlockConfig,
+    block_range: BlockRange,
+    _grpc_env: Arc<grpcio::Environment>,
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply fog/ledger/server/tests/router_integration.rs]: removed

File: fog/ledger/server/tests/router_integration.rs [hunk 5]
```diff
@@ -0,0 +1,572 @@
...
+            .iter()
+            .map(|x| x.address.to_string())
+            .collect(),
+    )
+}
+
+fn create_router(
+    test_config: &TestEnvironmentConfig,
+    blocks_config: &BlockConfig,
+    _grpc_env: Arc<grpcio::Environment>,
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply fog/ledger/server/tests/router_integration.rs]: removed

File: fog/ledger/server/tests/router_integration.rs [hunk 6]
```diff
@@ -0,0 +1,572 @@
...
+    );
+    store.start();
+
+    store
+}
+
+fn create_shard(
+    config: &ShardConfig,
+    _grpc_env: Arc<grpcio::Environment>,
+    _logger: Logger,
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply fog/ledger/server/tests/router_integration.rs]: _grpc_env was removed

File: fog/ledger/server/tests/router_integration.rs [hunk 7]
```diff
@@ -0,0 +1,571 @@
...
+    // Check that we can get all the key images from each shard
+    for i in 0..key_index {
+        let key = KeyImage::from(i);
+        let response = test_environment
+            .router_client
+            .check_key_images(&[key])
+            .await
+            .expect("check_key_images failed");
+        assert_eq!(response.results[0].key_image, key);
+        assert_eq!(response.results[0].status(), Ok(Some((i / 9) + 1)));
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply fog/ledger/server/tests/router_integration.rs]: refactored as suggested

File: fog/ledger/server/tests/router_integration.rs [hunk 8]
```diff
@@ -0,0 +1,571 @@
...
+    );
+
+    drop(test_environment.router_client);
+    test_environment.router.stop();
+    for shard in test_environment.shards {
+        let _ = shard.stop().await;
+    }
+    for mut store in test_environment.stores {
+        store.stop();
+    }
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply fog/ledger/server/tests/router_integration.rs]: implemented drop for `test_environment`

File: fog/ledger/server/tests/router_integration.rs [hunk 9]
```diff
@@ -0,0 +1,571 @@
...
+    let config = TestEnvironmentConfig {
+        router_address: free_sockaddr(),
+        router_admin_address: free_sockaddr(),
+        shards: shards_config,
+        omap_capacity: 1000,
+    };
+
+    let mut blocks_config = vec![];
+    let mut key_index = 0;
+    for _ in 1..5 {
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply fog/ledger/server/tests/router_integration.rs]: changed to `0..n`

File: fog/ledger/test_infra/src/lib.rs (hunk 2 above)
Comment (nick-mobilecoin) [reply fog/ledger/test_infra/src/lib.rs]: This was discussed offline. The purpose of the test is to test how the Router handles getting replies back from multiple shards. The shard proxy aren't tested directly as they will be implementation specific to the deployment.

**Comment by awygle** [comment]
`block_range` is there so we can configure stores incorrectly, which we'll need to close #2831 at the very least. So it will be used in a follow-up.

It's both shard overlap and store overlap, when stores are configured correctly stores and shards are configured to the same block range.

**Pull request merged**